### PR TITLE
Add check for release branch to make sure local branch is synced with remote

### PR DESCRIPTION
**Pull request by marsblkm**
- move branch check function to `ensureBranchUpToDate()` to make sure it can be done on any branch
- accomodate existing `ensureMainBranchUpToDate()`
- make sure release branch can be checked as well

Solve: https://github.com/sourcegraph/sourcegraph/issues/32783

## Test plan
@marsblkm tested locally and **manually** by doing the following:
1. Checkout to branch `3.39` and randomly commit something (to simulate the scenario)
2. Checkout to `main`
3. Merge `jm/check-local-and-remote-release-branch` into `main`.
4. Set up dry run mode and execute `yarn release release:create-candidate final`

image

### CODE REVIEW

**Comment by marsblkm** [comment]
@sourcegraph/contribution-reviewers

**Comment by caugustus-sourcegraph** [comment]
@sourcegraph/release-guild

File: dev/release/src/util.ts [hunk 1]
```diff
@@ -91,17 +108,21 @@ export function ensureMainBranchUpToDate(): void {
...
-    } else if (stdout.includes('Your branch is ahead')) {
-        console.log(`Your branch is ahead of the ${mainBranch} branch.`)
+    }
+}
+
+export function ensureReleaseBranchUpToDate(branch: string): void {
+    const currentBranch = execa.sync('git', ['rev-parse', '--abbrev-ref', 'HEAD']).stdout.trim()
+    execa.sync('git', ['checkout', branch])
+    const isReleaseBranchToDate = ensureBranchUpToDate(branch)
+    execa.sync('git', ['checkout', currentBranch])
```
Comment (keegancsmith) [inline dev/release/src/util.ts:120]: won't this break if you have local changes? eg when you tested this script and setup dry run stuff?

An alternative implementation is to run:
```shell
git fetch origin $branch
yours=$(git rev-parse $branch)
remote=$(git rev-parse origin/$branch)
# check if yours == remote
```

Comment (marsblkm) [reply dev/release/src/util.ts:120]: Hey this alternative implementation makes like much easier..

I do have a question tho.. do we care if there is any local changes? (changes that have not been staged...) for me when testing this I made some changes locally and I actually committed those changes.. so the script does not break. 

Now I guess for testing purpose, checking local changes is deliberately not introduced?..

Comment (marsblkm) [reply dev/release/src/util.ts]: Just update the code...
Also I manage to test it on `main` branch:
image

File: dev/release/src/util.ts [hunk 2]
```diff
@@ -81,8 +81,39 @@ export function changelogURL(version: string): string {
...
 
+function ensureBranchUpToDate(baseBranch: string, targetBranch: string): boolean {
+    const [behind, ahead] = execa.sync('git', ['rev-list', '--left-right', '--count', targetBranch + '...' + baseBranch]).stdout.split('\t')
+
+    if (behind === '0' && ahead === '0') {
+        return true
+    }
+
+    const countCommits = function (numberOfCommits: string, aheadOrBehind: string): string {
+        return numberOfCommits === '0' ? ''
```
Comment (keegancsmith) [inline dev/release/src/util.ts:92]: minor: you never pass in 0 as a number, so you can remove this case.

File: dev/release/src/util.ts [hunk 3]
```diff
@@ -91,17 +122,18 @@ export function ensureMainBranchUpToDate(): void {
...
     execa.sync('git', ['remote', 'update'], { stdout: 'ignore' })
-    const { stdout } = execa.sync('git', ['status', '-uno'])
-    if (stdout.includes('Your branch is behind')) {
-        console.log(
-            `Your branch is behind the ${mainBranch} branch. You should run \`git pull\` to update your ${mainBranch} branch.`
-        )
+    if (!ensureBranchUpToDate(mainBranch, remoteMainBranch)) {
         process.exit(1)
-    } else if (stdout.includes('Your branch is ahead')) {
-        console.log(`Your branch is ahead of the ${mainBranch} branch.`)
```
Comment (keegancsmith) [inline dev/release/src/util.ts:101]: it looks like in the old code we didn't exit if the branch was ahead. But the new code does? From what I can tell, we probably want the behaviour that your local branch reflects the origin, so this new behaviour makes sense.

**Review by keegancsmith — APPROVED** [review]
LGTM.

File: dev/release/src/util.ts (hunk 2 above)
Comment (marsblkm) [reply dev/release/src/util.ts]: Ah nice catch.. Just updated it.

**Comment by marsblkm** [comment]
Thanks for the approve @keegancsmith 

I don't think I have the rest of the workflow access. Hope someone in the team would not mind triggering them and get the changes merged eventually.

**Comment by unknwon** [comment]
For whoever is driving the merge, here is how to trigger a Buildkite build: https://handbook.sourcegraph.com/departments/product-engineering/engineering/process/external_contributions/

**Comment by keegancsmith** [comment]
Triggered a build, if its green and I forget to click merge please ping me :)

**Comment by marsblkm** [comment]
Just rerun the yarn prettier. build should be able to pass right now.

**Comment by keegancsmith** [comment]
triggered

**Comment by marsblkm** [comment]
Ah the build still does not work...
I cannot see the build log tho... any info that would like to provide there?

**Comment by keegancsmith** [comment]
@marsblkm you need to rebase onto the latest main branch

**Comment by marsblkm** [comment]
I just fetched the upstream main and rebase... Did I screw it up🥲

**Comment by keegancsmith** [comment]
@marsblkm yes. If you are unfamaliar with git it might be easiest if you just close this PR and recreate given the actual changes are quite small.

**Comment by marsblkm** [comment]
🥲 creating a new one now

**Pull request closed without merging**